File: ingest/model_plane.py
"""The model plane (TODO.ai-platform/05): the SMART Recommendation MODELS
join the retrieval.

The packages' machine content — the requirements' constraints, the
applicability rules, the acceptance criteria, the conformance tests, the
term definitions, the subject constraints, the characteristics, the state
machines — indexes ALONGSIDE the prose corpus. The index DERIVES from the
primmel packages (the single source of truth), never a hand-copied
extract: the smart repo's derive-model-plane.ts projects the packages into
committed bundles (browser/public/data/model-plane/*.json — byte-clean-
guarded by its ssot direction 4), and this module consumes those bundles
from the sibling smart checkout (SMART_REPO env; the estate's cross-repo
pattern — every upstream is a read-only sibling).

The freshness gate (the wave's must-pass leg): every bundle carries a
per-standard `source_hash` — a content hash over the primmel package. The
committed pins (ingest/model_plane_pins.json) record WHAT the index
derived from. `model-plane --check` compares: a package change moves the
hash, the gate exits 1, and the index is honestly stale until a re-index
(`python -m ingest.cli model-plane`) lands and refreshes the pins.
"""
from __future__ import annotations

import hashlib
import json
import os
import re
import sys
import tempfile
from pathlib import Path

from .config import ARTIFACTS

MODEL_CHUNKS_PATH = ARTIFACTS / "model_typed_chunks.jsonl"
MODEL_NODES_PATH = ARTIFACTS / "model_nodes.jsonl"
PINS_PATH = Path(__file__).resolve().parent / "model_plane_pins.json"

# Vectorize metadata caps (25 attrs/vector on the current plan); the
# model-node keys join the standard chunk keys well under the cap.
MAX_TEXT = 1800


BUNDLES_INDEX = "browser/public/data/model-plane/index.json"


def _sibling_candidates() -> list[Path]:
    """The estate's sibling layout: the checkout next to this repo, then
    the canonical ~/src/oimlsmart/smart."""
    return [Path(__file__).resolve().parents[2] / "smart", Path.home() / "src/oimlsmart/smart"]


def smart_repo_candidates() -> list[Path]:
    declared = os.environ.get("SMART_REPO", "").strip()
    cands = ([Path(declared)] if declared else []) + _sibling_candidates()
    # the sibling next to this repo can BE ~/src/oimlsmart/smart (the
    # estate layout) — name each real candidate once
    out: list[Path] = []
    seen: set[str] = set()
    for c in cands:
        key = str(c.resolve())
        if key not in seen:
            seen.add(key)
            out.append(c)
    return out


def _has_bundles(repo: Path) -> bool:
    return (repo / BUNDLES_INDEX).is_file()


def _git_branch(repo: Path) -> str:
    """The checkout's current branch, read straight from .git/HEAD (no
    subprocess; worktree .git files resolved). '' when unreadable."""
    git = repo / ".git"
    try:
        head = git / "HEAD"
        if git.is_file():
            # worktree/submodule: '.git' holds 'gitdir: <path>'
            line = git.read_text(encoding="utf-8").strip()
            if not line.startswith("gitdir:"):
                return ""
            gitdir = Path(line.split(":", 1)[1].strip())
            head = (repo / gitdir if not gitdir.is_absolute() else gitdir) / "HEAD"
        ref = head.read_text(encoding="utf-8").strip()
    except OSError:
        return ""
    if ref.startswith("ref: refs/heads/"):
        return ref.removeprefix("ref: refs/heads/")
    if re.fullmatch(r"[0-9a-f]{40}", ref):
        return f"detached @ {ref[:12]}"
    return ""


def _display(path: Path) -> str:
    home = str(Path.home())
    s = str(path)
    return "~" + s[len(home):] if s.startswith(home + os.sep) else s


def smart_repo(candidates: list[Path] | None = None) -> Path | None:
    """The smart checkout: SMART_REPO, else the estate's sibling layout —
    the first candidate carrying the committed model-plane bundles."""
    for c in (smart_repo_candidates() if candidates is None else candidates):
        if _has_bundles(c):
            return c
    return None


def smart_repo_diagnosis(candidates: list[Path] | None = None) -> str:
    """Why no candidate qualified — per-candidate detail so a checkout that
    EXISTS but lacks the bundles (parked on a legacy branch) is named with
    its branch instead of vanishing into a generic SKIP (the misread that
    produced oimlsmart/smart#252)."""
    cands = smart_repo_candidates() if candidates is None else candidates
    notes = []
    for c in cands:
        if not c.is_dir():
            notes.append(f"{_display(c)} does not exist")
            continue
        branch = _git_branch(c)
        where = f" (branch {branch})" if branch else ""
        notes.append(f"found {_display(c)}{where} but no {BUNDLES_INDEX} — not a v2 checkout carrying the committed bundles")
    if not notes:
        return "no candidates at all"
    if all(not c.is_dir() for c in cands):
        return "; ".join(notes)
    return "; ".join(notes) + " — declare SMART_REPO to a v2 checkout"


def load_bundles(repo: Path) -> dict[str, dict]:
    plane_dir = repo / "browser/public/data/model-plane"
    index = json.loads((plane_dir / "index.json").read_text(encoding="utf-8"))
    out: dict[str, dict] = {}
    for entry in index["standards"]:
        bundle = json.loads((plane_dir / entry["file"]).read_text(encoding="utf-8"))
        bundle["_plane"] = index.get("plane")
        out[bundle["standard"]] = bundle
    return out


def _urn_parts(base_urn: str) -> tuple[str, str, str]:
    # urn:oiml:pub:r:60:2021 → ("r", "60", "2021")
    m = re.match(r"^urn:oiml:pub:([rdbge]):(\d{1,3})(?::(\d{4}))?$", base_urn or "", re.I)
    if not m:
        return "", "", ""
    return m.group(1).lower(), m.group(2), m.group(3) or ""


def _clip(s: str | None, n: int = 600) -> str:
    s = (s or "").strip()
    return s if len(s) <= n else s[: n - 1].rstrip() + " …"


def _app_text(app: dict | None) -> str:
    if not app:
        return ""
    parts = []
    for dim, cond in app.items():
        if isinstance(cond, list):
            parts.append(f"{dim.replace('_', ' ')}: {', '.join(map(str, cond))}")
        elif isinstance(cond, dict) and isinstance(cond.get("values"), list):
            mode = cond.get("match", "any")
            parts.append(f"{dim.replace('_', ' ')} ({mode}): {', '.join(map(str, cond['values']))}")
    return "; ".join(parts)


def node_text(standard: str, bundle: dict, node: dict) -> str:
    """The grounding text for one model node — the node's OWN declared
    content, composed deterministically (nothing paraphrased)."""
    kind = node["kind"]
    label = bundle.get("label") or standard
    clause = node.get("clause") or {}
    head = f"[OIML SMART model — {kind.replace('_', ' ')}] {node.get('name') or node['id']} ({node['id']})"
    prov = label
    if clause.get("urn"):
        prov += f" · provenance {clause['urn']}"
    lines = [head, prov]
    if node.get("statement"):
        lines.append(f"Statement: {_clip(node['statement'])}")
    if node.get("definition"):
        lines.append(f"Definition: {_clip(node['definition'])}")
    if node.get("purpose"):
        lines.append(f"Purpose: {_clip(node['purpose'])}")
    if node.get("method"):
        lines.append(f"Method: {_clip(node['method'], 400)}")
    limit = node.get("limit") or {}
    if limit.get("expression"):
        lines.append(f"Machine limit (the constraint the platform's verdict engine evaluates): {limit['expression']}")
    if limit.get("accepts"):
        a = limit["accepts"]
        lines.append(f"Machine limit: {a.get('verdict')} {a.get('op')} {a.get('limit')} (canonical acceptance)")
    if node.get("check"):
        lines.append(f"Machine check: {node['check']}")
    if node.get("derive"):
        lines.append(f"Derivation: {node['derive']} (inputs: {', '.join(node.get('inputs') or [])})")
    app = _app_text(node.get("applicability"))
    if app:
        lines.append(f"Applicability: {app}")
    if node.get("binds_to"):
        lines.append(f"Binds to: {', '.join(node['binds_to'])}")
    if node.get("targets"):
        lines.append(f"Verifies requirements: {', '.join(node['targets'])}")
    if node.get("preconditions"):
        pcs = "; ".join(f"{p.get('id')}: {_clip(p.get('check') or ('state = ' + p['state'] if p.get('state') else ''), 120)}" for p in node["preconditions"])
        lines.append(f"Run-validity preconditions (a violation voids the run, never a fail): {pcs}")
    ac = node.get("acceptance_criteria") or {}
    if ac.get("description"):
        lines.append(f"Acceptance: {_clip(ac['description'], 300)}")
    if node.get("violation_meaning"):
        lines.append(f"Violation meaning (verbatim): {_clip(node['violation_meaning'], 300)} — on violation: {node.get('on_violation')}")
    if node.get("states"):
        trans = "; ".join(f"{t.get('from')}→{t.get('to')} ({t.get('action')})" for t in node.get("transitions") or [])
        lines.append(f"States: {', '.join(node['states'])}. Transitions: {_clip(trans, 400)}")
    if node.get("values"):
        vals = "; ".join(f"{v.get('id')}{' implies ' + ', '.join(v['implies']) if v.get('implies') else ''}" for v in node["values"])
        lines.append(f"Values: {_clip(vals, 400)}")
    if node.get("source_discrepancy"):
        sd = node["source_discrepancy"]
        lines.append(
            "DECLARED SOURCE DISCREPANCY (the model and the text disagree — always surface this and cite both): "
            + _clip(sd.get("summary"), 300)
            + f" Sources: {', '.join(sd.get('sources') or [])}. Resolution: {sd.get('resolution')} — {_clip(sd.get('rationale'), 300)}"
        )
    return "\n".join(lines)[:MAX_TEXT]


def node_chunk(standard: str, bundle: dict, node: dict) -> dict:
    doctype, doc_number, edition = _urn_parts(bundle.get("base_urn") or "")
    clause = node.get("clause") or {}
    cid = "m" + hashlib.sha1(f"{standard}|{node['id']}".encode()).hexdigest()[:16]
    text = node_text(standard, bundle, node)
    return {
        "id": cid,
        "doc_id": f"model:{standard}",
        "chunk_ref": f"{standard}{node['id']}",
        "text": text,
        "metadata": {
            "chunk_text": text,
            "doc_id": f"model:{standard}",
            "docidentifier": bundle.get("label") or standard,
            "doctype": doctype,
            "doc_number": doc_number,
            "edition": edition,
            "language": "en",
            "clause_anchor": clause.get("clause") or "model",
            "clause_title": f"{node['kind'].replace('_', ' ').title()} — {node.get('name') or node['id']}",
            "tier": "curated",
            "corpus": "smart-model",
            "status": "in-force",
            "superseded_by": "",
            "text_ref": f"model-plane/{standard}{node['id']}",
            "model_node": node["id"],
            "model_kind": node["kind"],
            "standard": standard,
        },
    }


def build(repo: Path, *, write_pins: bool = True) -> dict:
    """Emit the model-plane artifacts from the smart checkout's bundles."""
    bundles = load_bundles(repo)
    ARTIFACTS.mkdir(exist_ok=True)
    n_chunks = 0
    with MODEL_CHUNKS_PATH.open("w", encoding="utf-8") as f:
        for standard, bundle in sorted(bundles.items()):
            for node in bundle["nodes"]:
                f.write(json.dumps(node_chunk(standard, bundle, node), ensure_ascii=False) + "\n")
                n_chunks += 1
    n_nodes = 0
    with MODEL_NODES_PATH.open("w", encoding="utf-8") as f:
        for standard, bundle in sorted(bundles.items()):
            for node in bundle["nodes"]:
                content = json.dumps(node, ensure_ascii=False, sort_keys=True)
                clause = node.get("clause") or {}
                f.write(
                    json.dumps(
                        {
                            "standard": standard,
                            "node_id": node["id"],
                            "kind": node["kind"],
                            "name": node.get("name") or node["id"],
                            "clause_doc": clause.get("doc") or "",
                            "clause_ref": clause.get("clause") or "",
                            "content": content,
                            "content_hash": hashlib.sha256(content.encode()).hexdigest(),
                        },
                        ensure_ascii=False,
                    )
                    + "\n"
                )
                n_nodes += 1
    pins = {
        "plane": next(iter(bundles.values()), {}).get("_plane"),
        "standards": {
            s: {"source_hash": b["source_hash"], "node_count": len(b["nodes"])}
            for s, b in sorted(bundles.items())
        },
    }
    if write_pins:
        PINS_PATH.write_text(json.dumps(pins, indent=1) + "\n", encoding="utf-8")
    print(f"model-plane: {n_chunks} chunks + {n_nodes} nodes from {len(bundles)} standards")
    return pins


def check(repo: Path) -> int:
    """The freshness gate: the pinned source ≡ the smart checkout's current
    bundles. A package change moves a source_hash — the index is honestly
    stale until a re-index lands. Exit 1 on drift, 0 when current."""
    if not PINS_PATH.is_file():
        print("model-plane freshness: NO PINS — the model plane was never indexed (run python -m ingest.cli model-plane)")
        return 1
    pins = json.loads(PINS_PATH.read_text(encoding="utf-8"))
    bundles = load_bundles(repo)
    plane = next(iter(bundles.values()), {}).get("_plane")
    stale: list[str] = []
    if pins.get("plane") != plane:
        stale.append(f"the bundle shape moved ({pins.get('plane')} → {plane})")
    for standard, bundle in sorted(bundles.items()):
        pin = (pins.get("standards") or {}).get(standard)
        if not pin:
            stale.append(f"{standard}: never indexed")
            continue
        if pin.get("source_hash") != bundle["source_hash"]:
            stale.append(f"{standard}: the package moved (hash {pin.get('source_hash', '')[:12]}… → {bundle['source_hash'][:12]}…)")
        elif pin.get("node_count") != len(bundle["nodes"]):
            stale.append(f"{standard}: node count moved ({pin.get('node_count')} → {len(bundle['nodes'])})")
    for standard in (pins.get("standards") or {}):
        if standard not in bundles:
            stale.append(f"{standard}: indexed but no longer in the bundles")
    if stale:
        print("model-plane freshness: STALE — a package change re-indexes:")
        for s in stale:
            print(f"  ✗ {s}")
        print("re-index: .venv/bin/python -m ingest.cli model-plane  (then embed + upsert + the D1 apply)")
        return 1
    print(f"model-plane freshness: current ({len(bundles)} standards, pins ≡ bundles)")
    return 0


def _run_sql(path: Path) -> None:
    from .fts import _run_sql as fts_run_sql

    fts_run_sql(path)


def _esc(s) -> str:
    # bundle fields can arrive as lists (observed: two dead_load geometry
    # constraint names) — serialize anything non-string verbatim
    return (s if isinstance(s, str) else json.dumps(s, ensure_ascii=False)).replace("'", "''")


def apply() -> None:
    """Load model_nodes + model_plane_meta into D1 (the deterministic
    model-node store the ask path binds chips against). Idempotent:
    per-standard replace, meta upserted from the pins."""
    if not MODEL_NODES_PATH.is_file():
        raise SystemExit("run `model-plane` (build) first")
    pins = json.loads(PINS_PATH.read_text(encoding="utf-8")) if PINS_PATH.is_file() else {}
    rows = [json.loads(l) for l in MODEL_NODES_PATH.open(encoding="utf-8")]
    standards = sorted({r["standard"] for r in rows})
    statements = [f"DELETE FROM model_nodes WHERE standard IN ({', '.join(repr(s) for s in standards)});", "DELETE FROM model_plane_meta;"]
    for r in rows:
        statements.append(
            "INSERT OR REPLACE INTO model_nodes (standard, node_id, kind, name, clause_doc, clause_ref, content, content_hash) VALUES ("
            f"'{_esc(r['standard'])}','{_esc(r['node_id'])}','{_esc(r['kind'])}','{_esc(r['name'])}',"
            f"'{_esc(r['clause_doc'])}','{_esc(r['clause_ref'])}','{_esc(r['content'])}','{_esc(r['content_hash'])}');"
        )
    from datetime import datetime, timezone

    now = datetime.now(timezone.utc).isoformat()
    for standard in standards:
        pin = (pins.get("standards") or {}).get(standard) or {}
        statements.append(
            "INSERT OR REPLACE INTO model_plane_meta (standard, package, plane, source_hash, node_count, indexed_at) VALUES ("
            f"'{_esc(standard)}','{_esc(standard)}','{_esc(str(pins.get('plane') or ''))}',"
            f"'{_esc(str(pin.get('source_hash') or ''))}',{int(pin.get('node_count') or 0)},'{_esc(now)}');"
        )
    batch = 40
    for i in range(0, len(statements), batch):
        with tempfile.NamedTemporaryFile("w", suffix=".sql", delete=False) as tf:
            tf.write("\n".join(statements[i : i + batch]) + "\n")
            bpath = Path(tf.name)
        print(f"model-plane d1: {min(i + batch, len(statements))}/{len(statements)}", flush=True)
        _run_sql(bpath)
        bpath.unlink(missing_ok=True)
    print(f"model-plane d1: {len(rows)} nodes + {len(standards)} meta rows applied")


def main(argv: list[str]) -> int:
    write_pins = "--no-pins" not in argv
    do_check = "--check" in argv
    do_apply = "--apply" in argv
    repo = smart_repo()
    if repo is None:
        diag = smart_repo_diagnosis()
        if do_check:
            # the honest SKIP (exit 0 — the contract CI's ingest job rides
            # on) — but now it says WHY, per candidate
            print(f"model-plane freshness: SKIP — {diag}")
            return 0
        raise SystemExit(f"no smart checkout found: {diag}")
    if do_check:
        return check(repo)
    build(repo, write_pins=write_pins)
    if do_apply:
        apply()
    return 0


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: workers/worker_public/src/modelplane.ts
// The model plane (TODO.ai-platform/05) — the ask path's model-native
// grounding. The packages' machine content (the requirements' constraints,
// the applicability rules, the acceptance criteria, the term definitions)
// indexes alongside the prose corpus; this module is the deterministic
// half at ask time:
//
//   1. BIND — "this requirement" on a model surface (the platform's
//      entity chip) names the model node by its canonical id (the label
//      leads with it, e.g. "this requirement /req/metrological/mpe — …");
//      a question may name one too. The grammar is strict — a node id is
//      a machine shape, never fuzzy-matched.
//   2. GROUND — the bound node's own content (from the model_nodes D1
//      store, loaded from the smart repo's derived bundles) joins the
//      prompt as a structured grounding block: the constraint (quoted
//      verbatim), the applicability, the acceptance, the provenance, the
//      tests. A DECLARED source discrepancy (the model and the text
//      disagree) rides verbatim — the clause-drift doctrine's posture is
//      structural: the answer must surface it and cite both.
//   3. ECHO — context_applied.model names the bound node so the panel's
//      honest context line never invents the grounding.
//
// The verdict-explanation discipline is the smart repo's (its engine's
// trace seam — never reimplemented here): this module grounds in the
// model NODE (the static constraint + clause + acceptance); a live
// verdict's explanation is the platform's computation, quoted when the
// platform serves it — this service never recomputes one.

/** The strict model-node id grammar: the kind prefix + one or two
 *  slug segments (the packages' identifier shapes — /req/<class>/<id>,
 *  /conf/<class>/<id>, /term/<id>, /constraint/<id>, /characteristic/<id>,
 *  /state-machine/<id>, /dimension/<id>). */
import { P } from "./profile.ts";
const NODE_RE = /(?:^|[\s("'`])\/(req|conf|term|constraint|characteristic|state-machine|dimension)\/([a-z0-9][a-z0-9_-]*(?:\/[a-z0-9][a-z0-9_-]*)?)(?=[\s)"'`,;:.]|$)/i;

/** The first model-node id a text names (the declared chip label first,
 *  the question second — the caller orders). */
export function modelNodeRefIn(text: string | undefined | null): string | null {
  if (!text) return null;
  const m = text.match(NODE_RE);
  if (!m) return null;
  const id = `/${m[1]!.toLowerCase()}/${m[2]}`;
  return id.length <= 120 ? id : null;
}

/** The doc scope's publication family → the model plane's standard
 *  (doc_number "60" → oiml-r60). Only the four modeled Recommendations
 *  carry a plane; anything else resolves null (honest: no model to bind). */
export function standardForDocNumber(docNumber: string | undefined): string | null {
  if (!docNumber) return null;
  const models = P().sources?.models;
  if (!models?.standards?.length || !models?.standard_prefix) return null;
  return (models.standards as string[]).includes(docNumber) ? `${models.standard_prefix}${docNumber}` : null;
}

// ── the license tier (TODO.external-refs/08) ─────────────────────────────
// The profile's `sources.licensed` rows ({key, package, doc_number,
// title, edition}) are the deployment's declared licensed standards —
// the same keys the chunk metadata carries (`standard_key`) and the
// entitlement set arrives with (requestScope.standardKeysFrom validates
// against exactly this list).

/** The license entry for a package id (the model-plane standard id, e.g.
 *  `iec-60068-2-30`), or null when the package is public content. */
export function licensedEntryForPackage(packageId: string | undefined | null):
  | { key: string; package: string; doc_number?: string; title?: string; edition?: string }
  | null {
  if (!packageId) return null;
  return (P().sources?.licensed ?? []).find((l: any) => l.package === packageId) ?? null;
}

/** The license entry for a doc number (the question's named publication,
 *  e.g. "60068-2-30"), or null when unnamed/public. */
export function licensedEntryForDocNumber(docNumber: string | undefined | null):
  | { key: string; package: string; doc_number?: string; title?: string; edition?: string }
  | null {
  if (!docNumber) return null;
  return (P().sources?.licensed ?? []).find((l: any) => String(l.doc_number ?? "") === docNumber) ?? null;
}

/** The per-question license boundary note: composed ONLY when the
 *  question's named/understood publication is licensed AND the caller's
 *  entitlement set does not carry its key. The honesty posture, made
 *  structural: name the standard, say the organization's license does
 *  not cover its text, keep every procedural claim out, point at the
 *  declare flow. Citation-level metadata (title, edition, the invoking
 *  clause the RECs publicly name) stays answerable from the public
 *  passages already in context — the note instructs exactly that. */
export function licenseBoundaryNote(
  docNumber: string | undefined | null,
  standardKeys: ReadonlySet<string> | null | undefined,
): string | undefined {
  const entry = licensedEntryForDocNumber(docNumber);
  if (!entry || (standardKeys && standardKeys.has(entry.key))) return undefined;
  const pointer = P().prompts?.vars?.license_declare_pointer;
  return (
    `License boundary — the question is about ${licenseBoundaryName(entry)}, a licensed publication` +
    ` (entitlement key ${entry.key}). The caller's organization license does not cover its text, so no passage of it was retrieved` +
    ` and NONE of its procedural content (steps, parameters, severities, limits) may be stated, paraphrased or recalled from memory.` +
    ` You MAY answer at the citation level: name the standard and edition, and cite the invoking clause from the PUBLIC passages in context` +
    ` (the Recommendation's own applicability and normative references are public and stay answerable).` +
    ` Then say the organization's license does not cover the standard's text` +
    (pointer ? ` and point to the declare flow: ${pointer}.` : ".")
  );
}

function licenseBoundaryName(entry: { title?: string; edition?: string; package: string; doc_number?: string }): string {
  const id = entry.doc_number ? ` ${entry.doc_number}` : ` ${entry.package}`;
  return `${entry.title ?? "standard"}${entry.edition ? ` (${entry.edition})` : ""} —${id}`;
}

/** The deterministic boundary answer for the zero-passage case: the
 *  question's licensed publication has nothing to show an unentitled
 *  caller — name the standard, state the boundary, point at the declare
 *  flow. Undefined when the question is not the licensed case (the plain
 *  refusal applies). */
export function licenseBoundaryRefusal(
  docNumber: string | undefined | null,
  standardKeys: ReadonlySet<string> | null | undefined,
): string | undefined {
  const entry = licensedEntryForDocNumber(docNumber);
  if (!entry || (standardKeys && standardKeys.has(entry.key))) return undefined;
  const pointer = P().prompts?.vars?.license_declare_pointer;
  return (
    `${licenseBoundaryName(entry)} is a licensed publication and your organization's license does not cover its text, ` +
    `so I can't quote or summarize its procedure. I can answer at the citation level — the standard's title and edition, ` +
    `and the clause your Recommendation invokes — and the public ${P().publisher.name} content in full.` +
    (pointer ? ` To unlock the full text, an org admin can declare the license under ${pointer}.` : "")
  );
}

export interface BoundModelNode {
  standard: string;
  node_id: string;
  kind: string;
  name: string;
  clause: { doc: string; ref: string; urn: string } | null;
  /** The node's bundle projection (verbatim JSON). */
  content: any;
  /** True when the node's package is licensed and the caller's
   *  entitlement set lacks the key (TODO.external-refs/08): the citation
   *  and the echo stay (metadata), but the grounding block and the
   *  verdict engine are withheld — no licensed machine content enters
   *  the prompt. */
  gated?: boolean;
}

async function fetchNode(env: any, standard: string, nodeId: string): Promise<BoundModelNode | null> {
  try {
    const row = await env.DB.prepare(
      "SELECT standard, node_id, kind, name, clause_doc, clause_ref, content FROM model_nodes WHERE standard = ?1 AND node_id = ?2",
    )
      .bind(standard, nodeId)
      .first();
    if (!row) return null;
    const content = JSON.parse(String(row.content));
    const clause =
      row.clause_doc && row.clause_ref
        ? { doc: String(row.clause_doc), ref: String(row.clause_ref), urn: `${row.clause_doc}#clause-${row.clause_ref}` }
        : row.clause_doc
          ? { doc: String(row.clause_doc), ref: "", urn: String(row.clause_doc) }
          : null;
    return {
      standard: String(row.standard),
      node_id: String(row.node_id),
      kind: String(row.kind),
      name: String(row.name ?? row.node_id),
      clause,
      content,
    };
  } catch {
    // the table predates the migration on this deployment, or D1 hiccuped:
    // the model plane degrades honestly (retrieval still carries the model
    // chunks) — never a hard failure of the ask.
    return null;
  }
}

/** Bind the ask's model node: the declared entity label's id wins (the
 *  model-aware chip), then a node id the question names. The standard
 *  comes from the declared doc scope when it carries one; without a scope
 *  the node binds only when it exists in EXACTLY ONE indexed standard —
 *  ambiguity is refused honestly (retrieval still surfaces the chunks).
 *  A licensed package binds GATED for an unentitled caller (metadata
 *  only — the grounding block and the verdict engine are the ask path's
 *  to withhold). */
export async function bindModelNode(
  env: any,
  opts: { label?: string; query: string; standard?: string | null; standardKeys?: ReadonlySet<string> | null },
): Promise<BoundModelNode | null> {
  const nodeId = modelNodeRefIn(opts.label) ?? modelNodeRefIn(opts.query);
  if (!nodeId) return null;
  const gate = (node: BoundModelNode | null): BoundModelNode | null => {
    if (!node) return null;
    const entry = licensedEntryForPackage(node.standard);
    return entry && !(opts.standardKeys?.has(entry.key) ?? false) ? { ...node, gated: true, content: {} } : node;
  };
  if (opts.standard) return gate(await fetchNode(env, opts.standard, nodeId));
  try {
    const rows = await env.DB.prepare("SELECT standard FROM model_nodes WHERE node_id = ?1 LIMIT 2").bind(nodeId).all();
    const standards = (rows?.results ?? []).map((r: any) => String(r.standard));
    if (standards.length === 1) return gate(await fetchNode(env, standards[0]!, nodeId));
    return null; // zero (not indexed) or ambiguous (several standards) — no silent pick
  } catch {
    return null;
  }
}

// ── the grounding block (the node's own content, composed) ───────────

function clip(s: unknown, n = 500): string {
  const t = String(s ?? "").trim();
  return t.length <= n ? t : t.slice(0, n - 1).trimEnd() + " …";
}

function applicabilityText(app: any): string {
  if (!app || typeof app !== "object") return "";
  const parts: string[] = [];
  for (const [dim, cond] of Object.entries(app)) {
    if (Array.isArray(cond)) parts.push(`${dim.replace(/_/g, " ")}: ${cond.join(", ")}`);
    else if (cond && typeof cond === "object" && Array.isArray((cond as any).values)) {
      parts.push(`${dim.replace(/_/g, " ")} (${(cond as any).match ?? "any"}): ${(cond as any).values.join(", ")}`);
    }
  }
  return parts.join("; ");
}

/** The structured grounding block for the prompt — every line is the
 *  node's own declared content (the bundle projection), never a model
 *  paraphrase. The discrepancy block, when the model declares one, is the
 *  model/prose disagreement posture made structural. */
export function modelGroundingBlock(node: BoundModelNode): string {
  const c = node.content ?? {};
  const lines: string[] = [];
  lines.push(
    P().prompts.vars.model_grounding_intro ?? "Model grounding — the model plane's own statement:",
  );
  lines.push(`Node: ${node.node_id} (${node.kind.replace(/_/g, " ")}) — ${node.name} [${node.standard}]`);
  if (node.clause) lines.push(`Provenance: ${node.clause.urn}`);
  if (c.statement) lines.push(`Statement: ${clip(c.statement)}`);
  if (c.definition) lines.push(`Definition: ${clip(c.definition)}`);
  if (c.purpose) lines.push(`Purpose: ${clip(c.purpose)}`);
  const limit = c.limit ?? {};
  if (limit.expression) lines.push(`Machine limit (the constraint the platform's verdict engine evaluates — quote it verbatim): ${limit.expression}`);
  if (limit.accepts?.verdict) lines.push(`Machine limit: ${limit.accepts.verdict} ${limit.accepts.op} ${limit.accepts.limit} (the canonical acceptance chain)`);
  if (c.check) lines.push(`Machine check: ${c.check}`);
  if (c.derive) lines.push(`Derivation: ${c.derive}${Array.isArray(c.inputs) ? ` (inputs: ${c.inputs.join(", ")})` : ""}`);
  const app = applicabilityText(c.applicability);
  const scopeApp = applicabilityText(c.scope_applicability);
  if (app || scopeApp) lines.push(`Applicability: ${[scopeApp, app].filter(Boolean).join("; ")}`);
  if (Array.isArray(c.binds_to) && c.binds_to.length) lines.push(`Binds to: ${c.binds_to.join(", ")}`);
  if (Array.isArray(c.targets) && c.targets.length) lines.push(`Verifies requirements: ${c.targets.join(", ")}`);
  if (Array.isArray(c.preconditions) && c.preconditions.length) {
    const pcs = c.preconditions.map((p: any) => `${p.id}: ${clip(p.check ?? (p.state ? `state = ${p.state}` : ""), 120)}`).join("; ");
    lines.push(`Run-validity preconditions (a violation voids the run — invalid, never a fail): ${pcs}`);
  }
  if (c.acceptance_criteria?.description) lines.push(`Acceptance: ${clip(c.acceptance_criteria.description, 300)}`);
  if (c.violation_meaning) lines.push(`Violation meaning (verbatim): ${clip(c.violation_meaning, 300)} — on violation: ${c.on_violation ?? "invalid"}`);
  if (Array.isArray(c.values) && c.values.length) {
    lines.push(`Values: ${c.values.map((v: any) => `${v.id}${v.implies?.length ? ` (implies ${v.implies.join(", ")})` : ""}`).join("; ")}`);
  }
  if (c.source_discrepancy) {
    const sd = c.source_discrepancy;
    lines.push(
      `DECLARED SOURCE DISCREPANCY — the model and the text disagree; you MUST surface this and cite both: ${clip(sd.summary, 300)} ` +
        `Sources: ${(sd.sources ?? []).join(" and ")}. The model ${sd.resolution === "follows_clause_x" ? "follows one side" : "records the conflict without resolving it"}: ${clip(sd.rationale, 300)}`,
    );
  }
  lines.push(
    "Rules for this answer: the machine facts (the constraint, the applicability, the acceptance, the provenance) come from THIS node — quote the machine limit verbatim, never invent one the node does not carry. If this model content and a prose passage disagree — including a passage from a different edition — say so explicitly and cite both (this node and the prose clause).",
  );
  return lines.join("\n");
}

/** The citation the panel renders for the bound node (the model plane is
 *  a first-class corpus: the citation names the node + its clause). */
export function modelCitation(node: BoundModelNode) {
  return {
    doc_id: `model:${node.standard}`,
    docidentifier: `${P().publisher.name} SMART model (${(() => {
      const prefix = P().sources?.models?.standard_prefix ?? "";
      const letter = prefix.replace(/^.*-/, "").toUpperCase();
      return String(node.standard).replace(new RegExp(`^${prefix}`, "i"), `${letter} `);
    })()})`,
    edition: "",
    language: "en",
    clause_anchor: node.clause?.ref || "model",
    clause_title: `${node.kind.replace(/_/g, " ")} — ${node.name} (${node.node_id})`,
    status: "in-force",
    corpus: "smart-model",
    url: undefined,
    snippet: `${node.node_id}${node.clause ? ` · ${node.clause.urn}` : ""}${node.content?.statement ? ` — ${clip(node.content.statement, 240)}` : ""}`,
    score: 1,
  };
}

/** The context_applied echo's model block — the honest context line's
 *  grounding record. */
export function modelEcho(node: BoundModelNode) {
  return {
    node_id: node.node_id.slice(0, 120),
    kind: node.kind.slice(0, 40),
    standard: node.standard.slice(0, 40),
    ...(node.clause?.urn ? { clause: node.clause.urn.slice(0, 120) } : {}),
  };
}

/** The per-corpus guidance note (config.ts's DATASETS pattern — every
 *  retrieved model-plane chunk carries it, chip or no chip). */
export function modelCorpusNote(): string {
  return P().prompts.vars.model_passage_note ?? "";
}
